Core/Src/can_handler.c

```diff
@@ -0,0 +1,158 @@
+/**
+ * @file can_handler.c
+ * @author Hamza Iqbal
+ * @brief Source file for CAN handler.  
+ * @version 0.1
+ * @date 2023-09-22
+ * 
+ * @copyright Copyright (c) 2023
+ * 
+ */
+
+#include <stdlib.h>
+#include "can.h"
+#include "can_config.h"
+
+/* Initializes the CAN handler */
+void can_handler_init()
+{
+    // Assign memory to the queues
+    can1_incoming = malloc(sizeof(msg_queue));
+    can2_incoming = malloc(sizeof(msg_queue));
+    can1_outgoing = malloc(sizeof(msg_queue));
+    can2_outgoing = malloc(sizeof(msg_queue));
+
+    initializeHashMap(&callback_map);
+
+    for(int i = 0; i < NUM_CALLBACKS; i++)
+    {
+        insertFunction(&calllback_map, can_callbacks[i].messageID, can_callbacks[i].function);
+    }
+
+}
+
+/* Function to add a node to the message queue it is automatically called in the interrupt triggered callback */
+static void enqueue(struct msg_queue* queue, can_msg_t msg)
+{
+  struct node *new_node = malloc(sizeof(struct node));
+  new_node->msg = msg;
+  new_node->next = NULL;
+
+  if (queue->head == NULL)
+  {
+    queue->head = new_node;
+    queue->tail = new_node;
+  }
+  else
+  {
+    queue->tail->next = new_node;
+    queue->tail = new_node;
+  }
+}
+
+/* Removes and returns the front node of the queue */
+static can_msg_t* dequeue(struct msg_queue* queue)
+{
+    if (queue->head == NULL)
+    {
+        return NULL;
+    }
+
+    can_msg_t *msg = malloc(sizeof(can_msg_t));
+    *msg = queue->head->msg;
+    struct node *old_head = queue->head;
+    queue->head = queue->head->next;
+    free(old_head);
+
+    return msg;
+}
+
+/* Retrieves the message at the front of the queue and dequeues */
+can_msg_t *can_get_message(uint8_t line)
+{
+    can_msg_t *message;
+    switch(line)
+    {
+        case CAN_LINE_1:
+            message = dequeue(can1_incoming);
```

Comment:
So how do CAN messages actually get into the queue? I think you need to link the an interrupt handler to the CAN interrupt via the CAN driver we made (i.e. pass the pointer to the function to push a CAN message to a queue on init of CAN)